### PR TITLE
Implement cred store pal and bug fixes 0.1.1

### PR DESCRIPTION
**Pull request by marcschier**

### CODE REVIEW

File: api/csharp/samples/http/ReverseProxy.cs [hunk 1]
```diff
@@ -0,0 +1,704 @@
+﻿// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+namespace Microsoft.Azure.Devices.Proxy.Samples {
```
Comment (marcschier) [inline api/csharp/samples/http/ReverseProxy.cs]: Rename of sample

File: src/prx_ns.c [hunk 2]
```diff
@@ -1374,32 +1390,6 @@ static int32_t prx_ns_iot_hub_twin_entry_get_links(
...
-    void* context,
-    prx_ns_result_t** routes
-)
-{
-    (void)context;
-    (void)routes;
-    return er_not_found;
-}
-
-//
```
Comment (barnstee) [inline src/prx_ns.c]: Why did you remove routes?

Comment (marcschier) [inline src/prx_ns.c]: Was never implemented, it is time to remove unused code.

File: src/io_cs.c [hunk 3]
```diff
@@ -178,6 +183,139 @@ static int32_t io_cs_validate(
...
+            break;
+
+        // Success, delete original key
+        if (0 != STRING_compare(key_name, key_handle))
+            cs->entries[io_cs_entry_shared_access_key_handle] = key_handle;
+        else // No need to keep it - it is a unique well known key-name
+            STRING_delete(key_handle);
+
+        // Clear key value. 
+        // TODO: Need secure storage
```
Comment (marcschier) [inline src/io_cs.c]: Ha, I must have been someplace else when I wrote this comment.  I meant secure **memory**, i.e. non-pageable.  I do not think it is a hi prio threat on import path, but we might want to add a backlog item.

Comment (marcschier) [inline src/io_cs.c]: Ha, I must have been someplace else when I wrote this comment.  I meant secure **memory**, i.e. non-pageable.  I do not think it is a hi prio threat on import path, but we might want to add a backlog item.

**Comment by JuergenKosel** [comment]
Hello,

it looks like the database file could not be created anymore:

```
proxyd -i -n $DEVICENAME -C connectstring -D azure --only-websocket
Error: Time:Mon Mar 27 12:21:38 2017 File:iot-gateway-proxy/src/pal/pal_cred_stub.c Func:pal_cred_init Line:20 Pal does not provide credential storage support. Ensure you properly safeguard all secrets on this device!
Info: A OS operation resulted in error 2 (No such file or directory)
Info: A OS operation resulted in error 2 (No such file or directory)
ERROR: Failed to create local registry. 

```

File: api/csharp/samples/http/ReverseProxy.cs [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+        /// <param name="rewrite"></param>
+        /// <returns></returns>
+        protected async Task ServeAsync(HttpContext context, Uri uri,
+            Func<string, string> rewrite) {
+            // Connect a proxy socket to http endpoint
+
+            TcpClient client = null;
+            Stream stream = null;
+            try {
+                var target = uri.ToString(); // $"{uri.Scheme}://{uri.Host}{uri.PathAndQuery}";
```
Comment (mregen) [inline api/csharp/samples/http/ReverseProxy.cs]: what does it mean?

File: api/csharp/samples/http/ReverseProxy.cs [hunk 5]
```diff
@@ -0,0 +1,704 @@
...
+                    Uri redirect;
+                    if (!Uri.TryCreate(target, UriKind.RelativeOrAbsolute, out redirect) ||
+                        redirect.IsAbsoluteUri) {
+                        // if it is absolute, reconnect to the new host
+                        uri = redirect;
+                        target = uri.ToString();
+                    }
+                    else {
+                        //   // if target is relative, host is the same - see if we can reuse...
+                        //   if (context.Response.Headers["Connection"] == "keep-alive") {
```
Comment (mregen) [inline api/csharp/samples/http/ReverseProxy.cs]: tbd?

File: api/csharp/samples/http/ReverseProxy.cs [hunk 6]
```diff
@@ -0,0 +1,704 @@
...
+                        //       continue;
+                        //   }
+                    }
+
+                    stream.Dispose();
+                    client.Dispose();
+
+                    client = null;
```
Comment (mregen) [inline api/csharp/samples/http/ReverseProxy.cs]: better stream = null?

File: api/csharp/samples/http/ReverseProxy.cs [hunk 7]
```diff
@@ -0,0 +1,704 @@
...
+        /// <summary>
+        /// Connect a stream to a uri through the proxy
+        /// </summary>
+        /// <param name="client"></param>
+        /// <param name="uri"></param>
+        /// <returns></returns>
+        protected async Task<Stream> ConnectAsync(TcpClient client, Uri uri) {
+            await client.ConnectAsync(new ProxySocketAddress(uri.Host, uri.Port)).ConfigureAwait(false);
+            var stream = (Stream)client.GetStream();
+            if (!uri.Scheme.EndsWith("s")) {  // Crude way to find if we need to connect via ssl...
```
Comment (mregen) [inline api/csharp/samples/http/ReverseProxy.cs]: indeed!

**Comment by marcschier** [comment]
This is indeed broken.  Thanks for catching this @JuergenKosel.

**Comment by JuergenKosel** [comment]
Hello,

I have just cross checked, with the current master: There the creation of the database file seems to be already broken.

**Comment by marcschier** [comment]
@JuergenKosel, thanks for catching this, it is broken on Linux, since realpath returns error when the file does not exist.  I have fixed this in the pal_get_real_path implementation and will add the fix to this pull request.

**Comment by JuergenKosel** [comment]
@marcschier Do you know git bisect ( https://www.kernel.org/pub/software/scm/git/docs/git-bisect.html )?

git bisect shows, that the bug has been introduced by commit 0e4d9afebfcc064988bf83d058977ed387ddbb54
To get more useful results from git bisect, it is recommended to split changes into small logical steps, and commit each of them ( see section SEPARATE CHANGES of https://www.kernel.org/pub/software/scm/git/docs/gitworkflows.html ).
Just for the case, that some more difficult regression issue would be added in the future. :-)

Greetings
Juergen

File: bld/build.sh [hunk 8]
```diff
@@ -124,79 +123,79 @@ process_args ()
...
+            # Start as much parallel jobs as requested by the user.
+            # Until the load average equals the number of cores.
+            # Be verbose if something goes wrong
+            make -j$MAKE_PARALLEL_JOBS --load-average=`nproc` || make VERBOSE=1 || \
+                return 1
+            if [ $run_unittests == OFF ]; then
+                echo "Skipping unittests..."
+            else
+                #use doctored (-DPURIFY no-asm) openssl
+                export LD_LIBRARY_PATH=/usr/local/ssl/lib
```
Comment (JuergenKosel) [inline bld/build.sh]: You could set the environment variable LD_LIBRARY_PATH just for the invocation of ctest as follows:

`LD_LIBRARY_PATH=/usr/local/ssl/lib ctest -C "$c" --output-on-failure || return 1`

So you could keep the envrionment variable untouched for other programs called by the script.

Comment (marcschier) [inline bld/build.sh]: Ha, I did not know that.  I will change this.
